File: SADS/Tag/lib/Model/ElGamal.dart
import 'dart:math';
import 'package:beacon_simulator/Model/ECC.dart';
import 'package:beacon_simulator/Model/UseData.dart';

class ElGamal {
  late int numpoints;
  late ECC e;
  late UseData useData;
  late int a;
  late List g;
  late var h;

  ElGamal(int p, int a, int b, int numpoints) {
    this.numpoints = numpoints;
    this.e = ECC(p, a, b);
    this.useData = UseData();
  }

  List<dynamic> randomBase() {
    return this.useData.data[Random().nextInt(this.numpoints) + 0];
  }

  int createPrivateKey() {
    a = Random().nextInt(this.numpoints) + 0;
    return a;
  }

  multiply(x, y) {
    return e.addPoints(x, y);
  }

  exponentiate(x, n) {
    if (n > 0) {
      var temp = e.inversePoint(x);
      return e.multiplumOfPoint(temp, n);
    }
  }

  void keygen() {
    List<dynamic> base = randomBase();
    this.a = createPrivateKey();
    this.g = base;
    this.h = e.multiplumOfPoint(base, this.a);
  }

  decrypt(c, pk) {
    return multiply(c[1], exponentiate(c[0], pk));
  }
}

File: SADS/Server/ElGamal.py
import random
from ECC import ECC

class ElGamal():
    rand = random.SystemRandom()
    def __init__(self, p, a, b, numpoints, g, h):
        self.e = ECC(p=p, a=a, b=b)
        self.numpoints = numpoints
        self.g = g
        self.h = h

    def multiply(self,x,y):
        return self.e.addPoints(x,y)
   

    def exponentiate(self,x,n):
        if n < 0:
            x = self.e.inversePoint(x)
        return self.e.multiplumOfPoint(x,n)

    def encrypt(self, m):
        t = self.rand.randrange(0,self.numpoints)
        c1 = self.exponentiate(self.g,t)
        c2 = self.multiply(self.exponentiate(self.h,t),m)
        return [c1,c2]